File: cpp/0072_Edit_Distance.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cmath>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int len1 = word1.length(), len2 = word2.length();

        if (len1 == 0 || len2 == 0) return len1 + len2;

        vector<vector<int>> dp(len1 + 1, vector<int>(len2 + 1, 0));

        for (int i = 1; i <= len1; i++)
            dp[i][0] = i;

        for (int j = 1; j <= len2; j++)
            dp[0][j] = j;

        for (int i = 1; i <= len1; i++) {
            for (int j = 1; j <= len2; j++) {
                if (word1[i - 1] != word2[j - 1]) {
                    int tmp = min(dp[i - 1][j], dp[i][j - 1]);
                    dp[i][j] = min(tmp, dp[i - 1][j - 1]) + 1;
                } else {
                    dp[i][j] = dp[i - 1][j - 1];
                }
            }
        }

        return dp[len1][len2];
    }
};

int main(int argc, char *argv[]) {
    Solution solution;
    string word1, word2;

    // Input: word1 = "horse", word2 = "ros"
    // Output: 3
    word1 = "horse";
    word2 = "ros";
    cout << solution.minDistance(word1, word2) << endl;

    // Input: word1 = "intention", word2 = "execution"
    // Output: 5
    word1 = "intention";
    word2 = "execution";
    cout << solution.minDistance(word1, word2) << endl;

    return 0;
}

File: py/0072_Edit_Distance.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        len1, len2 = len(word1), len(word2)
        dp = [[0] * (len2 + 1) for _ in range(len1 + 1)]

        for i in range(1, len1 + 1):
            dp[i][0] = i

        for j in range(1, len2 + 1):
            dp[0][j] = j

        for i in range(1, len1 + 1):
            for j in range(1, len2 + 1):
                if word1[i - 1] == word2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1

        return dp[len1][len2]


if __name__ == "__main__":
    solution = Solution()

    # Input: word1 = "horse", word2 = "ros"
    # Output: 3
    print(solution.minDistance("horse", "ros"))

    # Input: word1 = "intention", word2 = "execution"
    # Output: 5
    print(solution.minDistance("intention", "execution"))

    sys.exit(0)
